### PR TITLE
[IDE] Set solution-specific variable types as interface types

### PR DESCRIPTION
**Pull request by ahoppen**
Setting the interface type of a variable, just to reset it to a null type is actually really gross. But quite a few methods further down in the generation of code completion results (such as USR generation) need to get a variable’s type and passing them along in a separate map would be really invasive. So this seems like the least bad solution to me.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

File: include/swift/IDE/TypeCheckCompletionCallback.h [hunk 1]
```diff
@@ -80,7 +80,31 @@ Type getPatternMatchType(const constraints::Solution &S, Expr *E);
...
+      if (!SolutionVarType.second->hasArchetype()) {
+        SolutionVarType.first->setInterfaceType(SolutionVarType.second);
+      } else {
+        SolutionVarType.first->setInterfaceType(ErrorType::get(SolutionVarType.second));
+      }
+    }
+  }
+
+  ~WithSolutionSpecificVarTypesRAII() {
+    for (auto Var : VarsWithSolutionSpecificTypes) {
```
Comment (rintaro) [inline include/swift/IDE/TypeCheckCompletionCallback.h:103]:
```suggestion
    for (auto *Var : VarsWithSolutionSpecificTypes) {
```

Comment (rintaro) [inline include/swift/IDE/TypeCheckCompletionCallback.h:103]:
```suggestion
    for (auto *Var : VarsWithSolutionSpecificTypes) {
```
to clarify this is a pointer.

File: include/swift/IDE/TypeCheckCompletionCallback.h [hunk 2]
```diff
@@ -80,7 +80,31 @@ Type getPatternMatchType(const constraints::Solution &S, Expr *E);
...
+      Var->setInterfaceType(Type());
```
Comment (rintaro) [inline include/swift/IDE/TypeCheckCompletionCallback.h:104]: Don't we need to reset this to the pre-RAII value? Or they are guaranteed to be null types? If so, could you add assertion in the constructor?

**Comment by ahoppen** [comment]
Note to self: The stress tester passed for this PR

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: include/swift/IDE/TypeCheckCompletionCallback.h [hunk 3]
```diff
@@ -78,9 +78,44 @@ Type getPatternMatchType(const constraints::Solution &S, Expr *E);
...
 /// were type-checked with the code completion expression.
-void getSolutionSpecificVarTypes(
-    const constraints::Solution &S,
-    llvm::SmallDenseMap<const VarDecl *, Type> &Result);
+void getSolutionSpecificVarTypes(const constraints::Solution &S,
+                                 llvm::SmallDenseMap<VarDecl *, Type> &Result);
+
+/// While this RAII is alive the interface types of the variables defined in
+/// \c SolutionSpecificVarTypes are reset to null types.
+/// Afterwards, their types are reset.
```
Comment (rintaro) [inline include/swift/IDE/TypeCheckCompletionCallback.h:86]:
```suggestion
/// While this RAII is alive the interface types of the variables defined in
/// \c SolutionSpecificVarTypes are temporality set to the type in the map.
/// Afterwards, their types are restored.
```

File: include/swift/IDE/ArgumentCompletion.h [hunk 4]
```diff
@@ -52,7 +52,7 @@ class ArgumentTypeCheckCompletionCallback : public TypeCheckCompletionCallback {
     /// Types of variables that were determined in the solution that produced
     /// this result. This in particular includes parameters of closures that
     /// were type-checked with the code completion expression.
-    llvm::SmallDenseMap<const VarDecl *, Type> SolutionSpecificVarTypes;
+    llvm::SmallDenseMap<VarDecl *, Type> SolutionSpecificVarTypes;
```
Comment (rintaro) [inline include/swift/IDE/ArgumentCompletion.h:55]: Non-`const` only for `InterfaceTypeRequest{VD}` in `WithSolutionSpecificVarTypesRAII::setInterfaceType`? If so, I would `const_cast` there, just like `ValueDecl::getInterfaceType()` ;)

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Review by rintaro — APPROVED** [review]

**Pull request merged**